### DISCUSSION TITLE
Do not embedd lemminx

### DISCUSSION THREAD

**Discussion by laeubi** [mickaelistria/eclipse-bluesky#1712] [Ideas]
I just faced an issue that I wanted to test a new lemminx snapshot build, and this was more complex than I thought.

1. https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/TIPS_and_FAQ.md didn't mentioned anything
2. I needed to debug the code to find out it has `lemminx` embedded and find the magic path (that will change depending on each deployment)
3. Then replace the jar by a manual copy to that location

What I thought about would it be not more convenient if lemminx would be not embedded in the jar itself? My idea would be to have the following adjustments:

1. make the lemminx-uber-jar a bundle that just exports all lemminx packages
2. let WWD require-bundle that one with a version range (e.g. excluding the next minor version)
3. the WWD can find out the JAR of the bundle and the just use that

This has the advantage that at least it is possible to update lemminx independent from WWD, the WWD bundle itself gets a lot "slimmer" (currently ~10 MB!) and possibly others can reuse that (e.g. extensions).

On the long run, one could even think about not using the uber-jar and simply collect lemminx and all its OSGi dependencies but thats a different topic.

**Comment by mickaelistria** [comment]
I tried it initially to make Wild Web Developer+LemMinX more OSGi-friendly and there are some challenges with it:
* In Maven world, there are not so fine-grain modularity so what works in the uber-jar (because Maven resolved and tested those dependencies) may fail in other deployments because of undocumented dependency ranges. In the current state, LemMinX is totally unable to support such "exotic" deployments.
* LemMinX uses ServiceLocator with classpath extensions to detect extension, and that is harder to support with OSGi
* Some LemMinX deps are/were not OSGi-ready

> make the lemminx-uber-jar a bundle that just exports all lemminx packages

That might be good; however I'm still unsure about how to enable extensibility there. If you can demonstrate extensibility can be made to work with such approach, then it is probably a very good step forward.

**Reply by laeubi** [reply]
> That might be good; however I'm still unsure about how to enable extensibility there. 

What exactly do you mean with extensibility? My main idea was to e.g start wit something really simple e.g. using

`org.eclipse.core.runtime.FileLocator.getBundleFileLocation(org.eclipse.core.runtime.Platform.getBundle("leminx-uber-jar"))`

in the `XMLLanguageServer` to get the jar location instead of using an embedded jar. 

Then one can install WWD+"leminx-uber-jar"(that currently is not a bundle!) bundles and also update them independently (as long as version ranges match of course), that would make thinks like [this](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/pull/1628) less urgent because users/products can choose a later version without any need to release WWD. This won't change much as it is of today.

> If you can demonstrate extensibility can be made to work with such approach, then it is probably a very good step forward

A next step could be to use `org.eclipse.core.runtime.Platform.getFragments(Bundle)` to collect fragments of the "leminx-uber-jar" bundle so one can easily add new extensions, as later on everything is put into a linear classpath there won't be much of an issue regarding finding them and again here then things can be updated more independently.

Then one can "strip out" things from the uber-jar that is well OSGIfied and use the Wiring API to find dependency bundles, again determine their Path by bundle and put them on the classpath of the server... again as its a flat classpath this won't harm but allow easier updates of dependencies (at least bugfix releases / Snapshots)

**Reply by mickaelistria** [reply]
> `org.eclipse.core.runtime.FileLocator.getBundleFileLocation(org.eclipse.core.runtime.Platform.getBundle("leminx-uber-jar"))`

Indeed, that seems like a very easy and powerful 1st iteration! Feel free to try it. Note that you'd most likely need to repackage the lemminx-uber-jar to add the proper MANIFEST; probably using bnd. If you do that, please name bundle with something like `org.eclipse.wildwebdeveloper.lemminxUberJar` and make sure the Bundle-Provider is `Eclipse Wild Web Developer`; unless you can ensure this can be provided and maintained by LemMinX.

**Reply by laeubi** [reply]
> unless you can ensure this can be provided and maintained by LemMinX.

I think that would be the best I'm currently investigate how this can be archived best but it seems these is no easy whay therefore I came up with a new Tycho mojo:
- https://github.com/eclipse-tycho/tycho/pull/4425

with that in place I'm able to enhance the uber-jar quite good already, it might has some issues if really used as OSGi artifact due to the heavy dependency chains but this can then be enhanced incrementally, e.g. one idea would be to import/export anything that is also available as OSGi, that way one can deploy it and already provided packages would be substituted.

**Reply by laeubi** [reply]
I have now created a draft PR at lemminx:
- https://github.com/eclipse-lemminx/lemminx/pull/1701

**Reply by mickaelistria** [reply]
Couldn't most of the change just take place in the .target of m2e?

**Reply by laeubi** [reply]
> Couldn't most of the change just take place in the .target of m2e?

This then wont allow an independent update of a lemminx build (so not much different from what it is today).

**Reply by mickaelistria** [reply]
The difference is that we wouldn't need to bump the version org org.eclipse.wildwebdeveloper.xml when trying to use a newer lemminx. That would already be a benefit.
But then, if you want lemminx to be "externally updatable", what would be the source p2 repo where new versions of lemminx come from?

**Reply by laeubi** [reply]
I think there are different cases

1. I build a local version / have lemminx in the workspace, this is/was the primary driver about my initial request as it was a bit tricky to find out how to get this working
2. there is a new lemminx release and I want to update it as a user independent from WWD, in such case e.g lemminx could provide a (maven based) update site

**Comment by laeubi** [comment]
@mickaelistria I have now created a PR here (needs approval from a maintainer):
- https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/pull/1741
